http/etag.ts

```diff
@@ -0,0 +1,196 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+/** Provides functions for dealing with and matching ETags, including
+ * {@linkcode calculate} to calculate an etag for a given entity,
+ * {@linkcode ifMatch} for validating if an ETag matches against a `If-Match`
+ * header and {@linkcode ifNoneMatch} for validating an Etag against an
+ * `If-None-Match` header.
+ *
+ * See further information on the `ETag` header on
+ * [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag).
+ *
+ * @module
+ */
+
+import { type DigestAlgorithm } from "../crypto/crypto.ts";
+import { toHashString } from "../crypto/to_hash_string.ts";
+import { createHash } from "../crypto/_util.ts";
+
+/** Just the part of `Deno.FileInfo` that is required to calculate an `ETag`,
+ * so partial or user generated file information can be passed. */
+export interface FileInfo {
+  mtime: Date | null;
+  size: number;
+}
+
+type Entity = string | Uint8Array | FileInfo;
+
+const DEFAULT_ALGORITHM = "FNV32A";
+const ENV_PERM_STATUS =
+  Deno.permissions.querySync?.({ name: "env", variable: "DENO_DEPLOYMENT_ID" })
+    .state ?? "granted"; // for deno deploy
+const DENO_DEPLOYMENT_ID = ENV_PERM_STATUS === "granted"
+  ? Deno.env.get("DENO_DEPLOYMENT_ID")
+  : undefined;
+const HASHED_DENO_DEPLOYMENT_ID = DENO_DEPLOYMENT_ID
+  ? createHash("FNV32A", DENO_DEPLOYMENT_ID).then((hash) => toHashString(hash))
+  : undefined;
+
+export interface ETagOptions {
+  /** A digest algorithm to use to calculate the etag. Defaults to
+   * `"FNV32A"`. */
+  algorithm?: DigestAlgorithm;
+
+  /** Override the default behavior of calculating the `ETag`, either forcing
+   * a tag to be labelled weak or not. */
+  weak?: boolean;
+}
+
+function isFileInfo(value: unknown): value is FileInfo {
+  return Boolean(
+    value && typeof value === "object" && "mtime" in value && "size" in value,
+  );
+}
+
+async function calcEntity(
```

Comment:
Imo the should be written out, `calculateEntity` instead of `calcEntity`

Reply:
They are internal and not exported. The term is not unclear.

Reply:
"calc" is widely-known to mean calculate. It's fine 🙂